### PR TITLE
Add socketOptions to RMQ connection for client and server

### PR DESCRIPTION
**Pull request by Upperfoot**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features) - https://github.com/nestjs/docs.nestjs.com/pull/244

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

You cannot specify the socket (connection) options detailed in https://www.squaremobius.net/amqp.node/channel_api.html#connect

Issue Number: N/A

## What is the new behavior?

You will now be able to specify the socket options (i.e. adding noDelay)

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

I came across this not being possible when I was looking at tuning our system, so decided to add it!

### CODE REVIEW

File: packages/microservices/client/client-rmq.ts [hunk 1]
```diff
@@ -49,6 +51,9 @@ export class ClientRMQ extends ClientProxy {
     this.queueOptions =
       this.getOptionsProp<RmqOptions>(this.options, 'queueOptions') ||
       RQM_DEFAULT_QUEUE_OPTIONS;
+    this.socketOptions =
+      this.getOptionsProp<RmqOptions>(this.options, 'socketOptions') ||
+      RQM_DEFAULT_SOCKET_OPTIONS;
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-rmq.ts]: do we have to use an additional class property?

File: packages/microservices/client/client-rmq.ts [hunk 2]
```diff
@@ -93,7 +98,7 @@ export class ClientRMQ extends ClientProxy {
   }
 
   public createClient<T = any>(): T {
-    return rqmPackage.connect(this.urls) as T;
+    return rqmPackage.connect(this.urls, this.socketOptions) as T;
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-rmq.ts]: maybe we could make `socketOptions` a local variable and just pluck this property inside the `createClient` method:
```
const socketOptions = this.getOptionsProp<RmqOptions>(this.options, 'socketOptions');
```

File: packages/microservices/constants.ts [hunk 3]
```diff
@@ -25,6 +25,7 @@ export const RQM_DEFAULT_QUEUE = 'default';
 export const RQM_DEFAULT_PREFETCH_COUNT = 0;
 export const RQM_DEFAULT_IS_GLOBAL_PREFETCH_COUNT = false;
 export const RQM_DEFAULT_QUEUE_OPTIONS = {};
+export const RQM_DEFAULT_SOCKET_OPTIONS = {};
```
Comment (kamilmysliwiec) [inline packages/microservices/constants.ts]: could be potentially removed? (just use `undefined` when not explicitly defined)

File: packages/microservices/server/server-rmq.ts [hunk 4]
```diff
@@ -43,6 +45,9 @@ export class ServerRMQ extends Server implements CustomTransportStrategy {
     this.queueOptions =
       this.getOptionsProp<RmqOptions>(this.options, 'queueOptions') ||
       RQM_DEFAULT_QUEUE_OPTIONS;
+    this.socketOptions =
+      this.getOptionsProp<RmqOptions>(this.options, 'socketOptions') ||
+      RQM_DEFAULT_SOCKET_OPTIONS;
```
Comment (kamilmysliwiec) [inline packages/microservices/server/server-rmq.ts]: everything same as in the client ^ class

**Comment by kamilmysliwiec** [comment]
hey @Upperfoot, thanks for the contribution! I left a few comments.

File: packages/microservices/client/client-rmq.ts (hunk 1 above)
Comment (Upperfoot) [reply packages/microservices/client/client-rmq.ts]: Thanks @kamilmysliwiec, your suggestion on it being called upon during the createClient is a much cleaner method, less properties flying around. I kind of wanted to keep parity with queueOptions but after looking at it further socketOptions wouldn't really be used outside of the context of createClient, good shout! 👍

**Comment by kamilmysliwiec** [comment]
Ping 🏓

**Comment by Upperfoot** [comment]
Pong! 🏓 @kamilmysliwiec

**Comment by kamilmysliwiec** [comment]
Thanks @Upperfoot :)

**Pull request merged**